### PR TITLE
Bug fix for total hits counts mismatch in hybrid query

### PR DESCRIPTION
**Pull request by vibrantvarun**
### Description
Solution Description
https://github.com/opensearch-project/neural-search/issues/756#issuecomment-2126278794

### Issues Resolved
[https://github.com/opensearch-project/neural-search/issues/756](https://github.com/opensearch-project/neural-search/issues/756)

### Check List
- [x] New functionality includes testing.
    - [x] All tests pass
- [ ] New functionality has been documented.
    - [ ] New functionality has javadoc added
- [x] Commits are signed as per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/test/java/org/opensearch/neuralsearch/processor/ScoreCombinationTechniqueTests.java [hunk 1]
```diff
@@ -27,7 +27,7 @@ public void testCombination_whenMultipleSubqueriesResultsAndDefaultMethod_thenSc
 
         final List<CompoundTopDocs> queryTopDocs = List.of(
             new CompoundTopDocs(
-                new TotalHits(3, TotalHits.Relation.EQUAL_TO),
+                new TotalHits(5, TotalHits.Relation.EQUAL_TO),
```
Comment (vibrantvarun) [inline src/test/java/org/opensearch/neuralsearch/processor/ScoreCombinationTechniqueTests.java:30]: Changing the count to 5 as this value will come from Query Phase. And in the data there are 5 unqiue scoredocs.

File: src/test/java/org/opensearch/neuralsearch/search/query/HybridQueryPhaseSearcherTests.java [hunk 2]
```diff
@@ -1025,7 +1025,9 @@ private static IndexMetadata getIndexMetadata() {
             RemoteStoreEnums.PathType.NAME,
             HASHED_PREFIX.name(),
             RemoteStoreEnums.PathHashAlgorithm.NAME,
-            RemoteStoreEnums.PathHashAlgorithm.FNV_1A_BASE64.name()
+            RemoteStoreEnums.PathHashAlgorithm.FNV_1A_BASE64.name(),
+            IndexMetadata.TRANSLOG_METADATA_KEY,
```
Comment (vibrantvarun) [inline src/test/java/org/opensearch/neuralsearch/search/query/HybridQueryPhaseSearcherTests.java:1029]: Added this setting to cater https://github.com/opensearch-project/OpenSearch/commit/76e2a5d9d70ab310033f33151fdaa0ab68c6537e#diff-240fea2643bc047fd12adfba28f9ba4e002bd0935e62b4f6fe49465526c3dc07R190

File: src/test/java/org/opensearch/neuralsearch/processor/ScoreCombinationTechniqueTests.java (hunk 1 above)
Comment (martin-gaievski) [reply src/test/java/org/opensearch/neuralsearch/processor/ScoreCombinationTechniqueTests.java:30]: Is this the only place in tests where we the check for totalHits was not correct? We may need to add more tests to check edge cases with totalHits

File: src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java [hunk 3]
```diff
@@ -80,16 +78,13 @@ private List<ScoreDoc> getCombinedScoreDocs(
...
     ) {
-        ScoreDoc[] finalScoreDocs = new ScoreDoc[maxHits];
-
         int shardId = compoundQueryTopDocs.getScoreDocs().get(0).shardIndex;
-        for (int j = 0; j < maxHits && j < sortedScores.size(); j++) {
-            int docId = sortedScores.get(j);
-            finalScoreDocs[j] = new ScoreDoc(docId, combinedNormalizedScoresByDocId.get(docId), shardId);
-        }
-        return Arrays.stream(finalScoreDocs).collect(Collectors.toList());
+        return sortedScores.stream()
```
[the remark this answers is not in the document]
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:84]: If I go with for loop then I need to initialize  ScoreDoc[] finalScoreDocs = new ScoreDoc[maxHits] array. Now the catch is maxHits is a long type and to create a array I need it to be integer. Typecasting long to int will not be the right thing.

File: src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java [hunk 4]
```diff
@@ -94,23 +97,23 @@ public void collect(int doc) throws IOException {
                 if (Objects.isNull(compoundQueryScorer)) {
                     throw new IllegalArgumentException("scorers are null for all sub-queries in hybrid query");
                 }
-
+                totalHits++;
```
[the remark this answers is not in the document]
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java:100]: Yes, great catch. I will do it

File: src/test/java/org/opensearch/neuralsearch/processor/ScoreCombinationTechniqueTests.java (hunk 1 above)
Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/processor/ScoreCombinationTechniqueTests.java:30]: Yes adding more.

File: src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:84]: Discussed offline, removing stream

[the remark this answers is not in the document]
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:84]: you should be able to create List of ScoreDoc objects right at the beginning, without intermediate array -> list conversion

File: src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java [hunk 5]
```diff
@@ -493,4 +497,70 @@ public void testCompoundScorer_whenHybridScorerIsTopLevelScorer_thenSuccessful()
...
+            .map(topdDoc -> topdDoc.scoreDocs)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+        Set<Integer> uniqueDocIds = new HashSet<>();
+        for (ScoreDoc[] scoreDocsArray : scoreDocs) {
+            uniqueDocIds.addAll(Arrays.stream(scoreDocsArray).map(scoreDoc -> scoreDoc.doc).collect(Collectors.toList()));
+        }
+        long maxTotalHits = uniqueDocIds.size();
+        assertNotEquals(maxTotalHits, totalHits);
+        assertEquals(4, totalHits);
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:561]: I assume this assert would fail without your change, right?

File: src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java [hunk 6]
```diff
@@ -493,4 +497,70 @@ public void testCompoundScorer_whenHybridScorerIsTopLevelScorer_thenSuccessful()
...
+        long totalHits = hybridTopScoreDocCollector.getTotalHits();
+        List<ScoreDoc[]> scoreDocs = topDocs.stream()
+            .map(topdDoc -> topdDoc.scoreDocs)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+        Set<Integer> uniqueDocIds = new HashSet<>();
+        for (ScoreDoc[] scoreDocsArray : scoreDocs) {
+            uniqueDocIds.addAll(Arrays.stream(scoreDocsArray).map(scoreDoc -> scoreDoc.doc).collect(Collectors.toList()));
+        }
+        long maxTotalHits = uniqueDocIds.size();
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:559]: it should be 1 based on the test logic, should we assert equals for it, something like:
```
assertEquals(1, maxTotalHits);
```

File: src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java (hunk 5 above)
Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:561]: Yes

File: src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java (hunk 6 above)
Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:559]: Sure I will add it tomorrow

Comment (martin-gaievski) [reply src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:559]: do you mean that each sub query will hav 1 unique hit, right? Otherwise it's not right as we're counting unique hits

**Comment by martin-gaievski** [comment]
LGTM, will approve once CI passed

File: src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java [hunk 7]
```diff
@@ -80,16 +78,15 @@ private List<ScoreDoc> getCombinedScoreDocs(
...
         final Map<Integer, Float> combinedNormalizedScoresByDocId,
         final List<Integer> sortedScores,
-        final int maxHits
+        final long maxHits
     ) {
-        ScoreDoc[] finalScoreDocs = new ScoreDoc[maxHits];
-
+        List<ScoreDoc> scoreDocs = new ArrayList<>();
         int shardId = compoundQueryTopDocs.getScoreDocs().get(0).shardIndex;
         for (int j = 0; j < maxHits && j < sortedScores.size(); j++) {
```
Comment (VijayanB) [inline src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:85]: shouldn't j be long too?

Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:85]: We have 2 conditions here, j<maxHits and sortedScores.size().  The sortedScores.size() condition will always return int. Therefore, It will be for j to be int.

**Comment by navneet1v** [comment]
> > @vibrantvarun good job on the PR. I would recommend adding an IT here or update the older IT to ensure that we are asserting on the totalhits in response.
> 
> @navneet1v I have added a dedicated unit test to verify the logic in this PR. Therefore, in the future if anything new added then there is always logic verification there in unit test.

I would check the IT for hybrid search and add the total hits check there(if not present) to rather than only unit test.

File: src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java [hunk 8]
```diff
@@ -135,7 +139,14 @@ public List<TopDocs> topDocs() {
...
-            .mapToObj(i -> topDocsPerQuery(0, Math.min(totalHits[i], compoundScores[i].size()), compoundScores[i], totalHits[i]))
+            .mapToObj(
+                i -> topDocsPerQuery(
+                    0,
+                    Math.min(collectedHitsPerSubQuery[i], compoundScores[i].size()),
+                    compoundScores[i],
+                    collectedHitsPerSubQuery[i]
+                )
+            )
             .collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (shatejas) [reply src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java]: Noted. I can imagine it being slow when you terminate the stream every now and then

**Review by navneet1v — APPROVED** [review]

File: src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (heemin32) [reply src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java]: @vibrantvarun Do we have a data showing how much does stream slow down the query performance against for loop?

[the remark this answers is not in the document]
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java]: @martin-gaievski has the flame graph for performance optimization. Martin Can you  share those data points here?
